src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1147,9 +1147,32 @@ async ValueTask<NpgsqlDataReader> ExecuteReaderAsync(CommandBehavior behavior, b
                 if (conn.TryGetBoundConnector(out var connector))
                 {
                     connector.StartUserAction(this);
+
+                    CancellationTokenSource? cts = null;
+                    CancellationTokenRegistration? registration = null;
+                    var finalCt = CancellationToken.None;
+                    var cancellationTimeout = connector.Settings.CancellationTimeout;
+
                     try
                     {
-                        using var _ = cancellationToken.Register(cmd => ((NpgsqlCommand)cmd!).Cancel(), this);
+                        if (cancellationToken.CanBeCanceled)
+                        {
+                            cts = new CancellationTokenSource();
```

Comment:
From a high-perf perspective, it isn't amazing to allocate a cts on every execution. Also, since it's referenced from the lambda below, the local is captured (closure) for further allocation.

Since only one command can be running at a given point on a connector, we can simply move the cts to the connector, and even recycle it as long as it hasn't fired (like we do in multiplexing/NpgsqlReadBuffer etc.).

Reply:
This is a great idea! High-perf is something I usually don't have to deal with, so your thoughts from high-perf perspective are really appreciated 😄

Reply:
Yeah, we try to be really really optimized on the fast paths of Npgsql. .NET is ranked quite high on the [TechEmpower benchmarks](https://www.techempower.com/benchmarks/#section=data-r19), and Npgsql is part of that. Allocations are one place where it's good to be very very strict to remove pressure on the garbage collector.